src/main/java/de/zalando/aruha/nakadi/service/subscription/StreamingContext.java

```diff
@@ -0,0 +1,170 @@
+        return zkClient;
+    }
+
+    public String getSessionId() {
+        return session.getId();
+    }
+
+    public KafkaClient getKafkaClient() {
+        return kafkaClient;
+    }
+
+    public SubscriptionOutput getOut() {
+        return out;
+    }
+
+    public long getKafkaPollTimeout() {
+        return kafkaPollTimeout;
+    }
+
+    @Override
+    public void stream() throws InterruptedException {
+        currentState = new State() {
+            @Override
+            public void onEnter() {
+            }
+        };
+
+        // Add first task - switch to starting state.
+        switchState(new StartingState());
+
+        while (null != currentState) {
+            final Runnable task = taskQueue.poll(1, TimeUnit.MINUTES);
+            try {
+                task.run();
+            } catch (final SubscriptionWrappedException ex) {
+                LOG.error("Failed to process task " + task + ", will rethrow original error", ex);
+                switchState(new CleanupState(ex.getSourceException()));
+            } catch (final RuntimeException ex) {
+                LOG.error("Failed to process task " + task + ", code carefully!", ex);
+                switchState(new CleanupState(ex));
+            }
+        }
+    }
+
+    public void switchState(final State newState) {
+        this.addTask(() -> {
+            if (currentState != null) {
+                LOG.info("Switching state from " + currentState.getClass().getSimpleName());
+                currentState.onExit();
+            } else {
+                LOG.info("Connection died, no new state switches available");
+                return;
+            }
+            currentState = newState;
+            if (null != currentState) {
+                LOG.info("Switching state to " + currentState.getClass().getSimpleName());
+                currentState.setContext(this);
+                currentState.onEnter();
+            } else {
+                LOG.info("No next state found, dying");
+            }
+        });
+    }
+
+    public void registerSession() {
+        LOG.info("Registering session " + session);
+        zkClient.registerSession(session);
+        // Install rebalance hook on client list change.
+        clientListChanges = zkClient.subscribeForSessionListChanges(() -> addTask(this::rebalance));
+
+        // Invoke rebalance directly
+        addTask(this::rebalance);
+    }
+
+    public void unregisterSession() {
+        LOG.info("Unregistering session " + session);
+        if (null != clientListChanges) {
+            try {
+                clientListChanges.cancel();
+            } finally {
+                this.clientListChanges = null;
+                zkClient.unregisterSession(session);
+            }
+        }
+    }
+
+    public boolean isInState(final State state) {
+        return currentState == state;
+    }
+
+    public void addTask(final Runnable task) {
+        taskQueue.offer(task);
```

Comment:
How do you decide which things to execute within this task queue and which things to execute without queue?